smithy-model/src/main/java/software/amazon/smithy/model/shapes/EnumShape.java

```diff
@@ -355,9 +388,33 @@ public Builder setMembersFromEnumTrait(EnumTrait trait, boolean synthesizeNames)
                 }
             }
 
+            boolean hasDuplicateMemberName = false;
+
+            // Check for synthesized name duplication
+            if (synthesizeNames) {
+                Set<String> names = new HashSet<>();
+                members.forEach(m -> names.add(m.getMemberName()));
+                hasDuplicateMemberName = names.size() != members.size();
+            }
+
+
+            for (int i = 0; i < members.size(); i++) {
+                MemberShape member = members.get(i);
+                if (hasDuplicateMemberName) {
```

Comment:
As this stands this change will penalize each name regardless of whether or not these  collide with another one. I think is better to just penalize those that do. For instance, if we have 

* `FOO`
* `FOO:BAR`
* `FOO=BAR`
* `BAR`

I think it would be better to get

* `FOO` → `FOO`
* `FOO:BAR` → `FOO_BAR_0`
* `FOO=BAR` → `FOO_BAR_1`
* `BAR` → `BAR`

Instead of

* `FOO` → `FOO_0`
* `FOO:BAR` → `FOO_BAR_1`
* `FOO=BAR` → `FOO_BAR_2`
* `BAR` → `BAR_3`